File: server/src/modules/message/message.controller.ts
import { Context } from "koa";
import { Result } from "@/core/result";
import { MessagePageDto, SendMessageDto } from "./message.dto";
import { messageService } from "./message.service";

/** 从 JWT 中取当前用户 ID */
function currentUserId(ctx: Context): number {
  return (ctx.state as any).user?.id as number;
}

class MessageController {
  async page(ctx: Context) {
    const parsed = MessagePageDto.safeParse(ctx.query);
    if (!parsed.success) {
      ctx.body = Result.error(parsed.error.issues?.[0]?.message ?? "参数错误");
      return;
    }
    const uid = currentUserId(ctx);
    const { list, total } = await messageService.page(parsed.data, uid);
    ctx.body = Result.page(list, total, parsed.data.page, parsed.data.pageSize);
  }

  async unreadCount(ctx: Context) {
    const uid = currentUserId(ctx);
    ctx.body = Result.success(await messageService.unreadCount(uid));
  }

  async markRead(ctx: Context) {
    const id = Number(ctx.params.id);
    if (!id || Number.isNaN(id)) {
      ctx.body = Result.error("参数错误");
      return;
    }
    const ok = await messageService.markRead(id, currentUserId(ctx));
    ctx.body = Result.success(ok, ok ? "已读" : "消息不存在或已读");
  }

  async markAllRead(ctx: Context) {
    const count = await messageService.markAllRead(currentUserId(ctx));
    ctx.body = Result.success(count, "已全部标记已读");
  }

  /**
   * 可选：通过 REST 发送一条站内消息（便于联调/后台运营）
   * - senderId 从 token 推导，避免伪造
   */
  async send(ctx: Context) {
    const parsed = SendMessageDto.safeParse(ctx.request.body);
    if (!parsed.success) {
      ctx.body = Result.error(parsed.error.issues?.[0]?.message ?? "参数错误");
      return;
    }
    const uid = currentUserId(ctx);
    const created = await messageService.sendRealTimeMessage(parsed.data, uid);
    ctx.body = Result.success(created, "发送成功");
  }
}

export const messageController = new MessageController();

File: java-server/src/main/java/com/todothink/modules/message/controller/MessageController.java
package com.todothink.modules.message.controller;

import com.todothink.core.result.Result;
import com.todothink.core.security.SecurityUtils;
import com.todothink.modules.message.dto.SendMessageRequest;
import com.todothink.modules.message.service.MessageService;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/messages")
public class MessageController {

    private final MessageService messageService;

    public MessageController(MessageService messageService) {
        this.messageService = messageService;
    }

    @GetMapping(value = {"", "/"})
    public Result<?> page(@RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int pageSize,
            @RequestParam(required = false) String isRead,
            @RequestParam(required = false) String messageType) {
        Boolean readFilter = null;
        if ("true".equalsIgnoreCase(isRead)) {
            readFilter = true;
        } else if ("false".equalsIgnoreCase(isRead)) {
            readFilter = false;
        }
        Map<String, Object> res = messageService.page(page, pageSize, readFilter, messageType,
                SecurityUtils.currentUserId());
        return Result.page((List<?>) res.get("list"), (Long) res.get("total"), page, pageSize);
    }

    @GetMapping("/unread-count")
    public Result<?> unreadCount() {
        return Result.success(messageService.unreadCount(SecurityUtils.currentUserId()));
    }

    @PostMapping("/{id}/read")
    public Result<?> markRead(@PathVariable Integer id) {
        boolean ok = messageService.markRead(id, SecurityUtils.currentUserId());
        return Result.success(ok, ok ? "已读" : "消息不存在或已读");
    }

    @PostMapping("/read-all")
    public Result<?> markAllRead() {
        int count = messageService.markAllRead(SecurityUtils.currentUserId());
        return Result.success(count, "已全部标记已读");
    }

    @PostMapping(value = {"", "/"})
    public Result<?> send(@Validated @RequestBody SendMessageRequest body) {
        Map<String, Object> created = messageService.sendRealTimeMessage(body, SecurityUtils.currentUserId());
        return Result.success(created, "发送成功");
    }
}
